include/zephyr/drivers/stepper.h

```diff
@@ -434,6 +437,29 @@ static inline int z_impl_stepper_enable_constant_velocity_mode(
 	return api->enable_constant_velocity_mode(dev, direction, value);
 }
 
+/**
+ * @brief Set the callback function to be called when a stepper event occurs
+ *
+ * @param dev pointer to the stepper motor controller instance
+ * @param callback Callback function to be called when a stepper event occurs
+ * passing NULL will disable the callback
+ *
+ * @retval -ENOSYS If not implemented by device driver
+ * @retval 0 Success
+ */
+__syscall int stepper_set_callback(const struct device *dev, stepper_event_callback_t callback);
```

Comment:
you need to add the `void *user_data` argument :) The user data is usually used to point to the context of whatever is setting the callback :)